In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data = StandardScaler().fit_transform(df_market_data[[
    "price_change_percentage_24h",
    "price_change_percentage_7d",
    "price_change_percentage_14d",
    "price_change_percentage_30d",
    "price_change_percentage_60d",
    "price_change_percentage_200d",
    "price_change_percentage_1y"]])

In [6]:
# Create a DataFrame with the scaled data
scaled_data_df = pd.DataFrame(scaled_data, columns=[
    "price_change_percentage_24h", 
    "price_change_percentage_7d",
    "price_change_percentage_14d",
    "price_change_percentage_30d",
    "price_change_percentage_60d",
    "price_change_percentage_200d",
    "price_change_percentage_1y"])

# Copy the crypto names from the original DataFrame
scaled_data_df["coin_id"]=df_market_data.index

# Set the coin_id column as index
scaled_data_df.set_index("coin_id", inplace=True)

# Display the scaled DataFrame
scaled_data_df.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

### Find the Best Value for k Using the Original Scaled DataFrame.

In [7]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [10]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model=KMeans(n_clusters=i)
    k_model.fit(scaled_data_df)
    inertia.append(k_model.inertia_)

In [11]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_data_df = pd.DataFrame(elbow_data)

elbow_data_df.head(10)

k     inertia
0   1  287.000000
1   2  198.571818
2   3  145.635748
3   4  105.399719
4   5   65.284985
5   6   64.281150
6   7   46.035250
7   8   37.527740
8   9   32.630543
9  10   28.324714

In [56]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_plot=elbow_data_df.hvplot.line(
    x="k",
    y ="inertia", 
    title = "Elbow Curve",
    xticks=k, 
    color="blue")
elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value of k, for the appropriate number of clusters for this data set, is 5. This is indiccated by the "elbow" shape in the line graph of k-value vs. inertia at the k-value of 5.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [14]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=5)

model

KMeans(n_clusters=5)

In [15]:
# Fit the K-Means model using the scaled DataFrame
model.fit(scaled_data_df)

model

KMeans(n_clusters=5)

In [16]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
spread_predictions = model.predict(scaled_data_df)

# Print the resulting array of cluster values.
print(spread_predictions)

[1 0 1 1 0 0 0 0 1 1 4 1 1 0 1 1 4 1 0 1 4 1 4 1 1 1 1 4 0 4 1 1 2 1 4 4 3
 4 1 1 4]


In [17]:
# Create a copy of the scaled DataFrame
spread_predictions_df = scaled_data_df.copy()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

In [18]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
spread_predictions_df["Cluster Predictions"]= spread_predictions

# Display the copy of the scaled DataFrame
spread_predictions_df.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  Cluster Predictions  
coin_id                                                           
bitcoin                           -0.251637                    1  
ethereum                          -0.199352                    0  
tether                            -0.282061                    1  
ripple                            -0.295546                    1  
bitcoin-cash                      -0.270317                    0  
binancecoin                       -0.225533                    0  
chainlink                         -0.018284                    0  
cardano                           -0.155428                    0  
litecoin                          -0.292351                    1  
bitcoin-cash-sv                   -0.206029                    1

In [65]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
cluster_plot=spread_predictions_df.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d", 
    by="Cluster Predictions",
    hover_cols="coin_id")
cluster_plot

:NdOverlay   [Cluster Predictions]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

### Optimize Clusters with Principal Component Analysis.

In [23]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [27]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
crypto_pca = pca.fit_transform(scaled_data_df)

# View the scaled PCA data
crypto_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [31]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
print(pca.explained_variance_ratio_)

sum(pca.explained_variance_ratio_)

[0.3719856  0.34700813 0.17603793]


0.8950316570309841

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total epxlained varianve of the principal compnonents in this instance is 0.895. This means that approximately 89.5% of the data was preserved while approximately 10.5% of the data was lost in the process of principal component analysis. 

In [36]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(
    crypto_pca,
    columns=["PCA1", "PCA2", "PCA3"])

# Copy the crypto names from the original scaled DataFrame
df_market_data_pca["coin_id"]=scaled_data_df.index

# Set the coin_id column as index
df_market_data_pca.set_index("coin_id", inplace=True)

# Display the scaled PCA DataFrame
df_market_data_pca.head(10)

PCA1      PCA2      PCA3
coin_id                                      
bitcoin         -0.600667  0.842760  0.461595
ethereum        -0.458261  0.458466  0.952877
tether          -0.433070 -0.168126 -0.641752
ripple          -0.471835 -0.222660 -0.479053
bitcoin-cash    -1.157800  2.041209  1.859715
binancecoin     -0.516534  1.388377  0.804071
chainlink       -0.450711  0.517699  2.846143
cardano         -0.345600  0.729439  1.478013
litecoin        -0.649468  0.432165  0.600303
bitcoin-cash-sv -0.759014 -0.201200 -0.217653

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [35]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [39]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    modelpca=KMeans(n_clusters=i)
    modelpca.fit(df_market_data_pca)
    inertia.append(modelpca.inertia_)

In [40]:
# Create a dictionary with the data to plot the Elbow curve
elbow_pca_data={"k":k, "inertia":inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_pca_df = pd.DataFrame(elbow_pca_data)

elbow_pca_df.head()

k     inertia
0  1  256.874086
1  2  168.811896
2  3   93.774626
3  4   49.665497
4  5   38.352251

In [57]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
pca_elbow_plot=elbow_pca_df.hvplot.line(
    x="k",
    y="inertia", 
    title="Elbow Curve", 
    xticks=k,
    color="red")
pca_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value of k, for the appropriate number of clusters for this data set, is 4. This is indiccated by the "elbow" shape in the line graph pf k-value vs. inertia at the k-value of 4.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes, this does differ from the k value found using the original data, whick was a value of 5. 

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [44]:
# Initialize the K-Means model using the best value for k
model_pca=KMeans(n_clusters=4)
model_pca

KMeans(n_clusters=4)

In [45]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_market_data_pca)

model_pca

KMeans(n_clusters=4)

In [46]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_predictions=model_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values.
print(pca_predictions)

[0 0 2 2 0 0 0 0 0 2 2 2 2 0 2 0 2 2 0 2 2 0 2 2 2 2 2 2 0 2 2 2 1 0 2 2 3
 2 2 2 2]


In [48]:
# Create a copy of the scaled PCA DataFrame
market_data_pca_predictions_df=df_market_data_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
market_data_pca_predictions_df["Cluster Predictions"]=pca_predictions

# Display the copy of the scaled PCA DataFrame
market_data_pca_predictions_df.head(10)

PCA1      PCA2      PCA3  Cluster Predictions
coin_id                                                           
bitcoin         -0.600667  0.842760  0.461595                    0
ethereum        -0.458261  0.458466  0.952877                    0
tether          -0.433070 -0.168126 -0.641752                    2
ripple          -0.471835 -0.222660 -0.479053                    2
bitcoin-cash    -1.157800  2.041209  1.859715                    0
binancecoin     -0.516534  1.388377  0.804071                    0
chainlink       -0.450711  0.517699  2.846143                    0
cardano         -0.345600  0.729439  1.478013                    0
litecoin        -0.649468  0.432165  0.600303                    0
bitcoin-cash-sv -0.759014 -0.201200 -0.217653                    2

In [66]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
pca_cluster_plot=market_data_pca_predictions_df.hvplot.scatter(
    x="PCA1", 
    y="PCA2",
    by="Cluster Predictions",
    hover_cols="coin_id")
pca_cluster_plot

:NdOverlay   [Cluster Predictions]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [64]:
# Composite plot to contrast the Elbow curves
composite_plot =elbow_plot*pca_elbow_plot

composite_plot

:Overlay
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [68]:
# Composite plot to contrast the clusters
composite_clusters=cluster_plot*pca_cluster_plot

composite_clusters

:Overlay
   .NdOverlay.I  :NdOverlay   [Cluster Predictions]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [Cluster Predictions]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 